tree_in_context/src/add_context.rs

```diff
@@ -0,0 +1,94 @@
+            // leaf nodes are constant, empty, and arg
+            Expr::Const(..) | Expr::Empty(..) | Expr::Arg(..) => {
+                in_context(current_ctx, self.clone())
+            }
+            // create new contexts for let, loop, and if
+            Expr::DoWhile(inputs, pred_and_body) => {
+                let new_inputs = inputs.add_context(current_ctx.clone());
+                let new_ctx = Assumption::InLoop(new_inputs.clone(), pred_and_body.clone());
+                RcExpr::new(Expr::DoWhile(
+                    new_inputs,
+                    pred_and_body.add_context(new_ctx),
+                ))
+            }
+            Expr::Let(inputs, body) => {
+                let new_inputs = inputs.add_context(current_ctx.clone());
+                let new_ctx = Assumption::InLet(new_inputs.clone());
+                RcExpr::new(Expr::Let(new_inputs, body.add_context(new_ctx)))
+            }
+            Expr::If(pred, then_case, else_calse) => {
+                let new_pred = pred.add_context(current_ctx.clone());
+                let then_ctx = Assumption::InIf(true, new_pred.clone());
+                let else_ctx = Assumption::InIf(false, new_pred.clone());
+                RcExpr::new(Expr::If(
+                    new_pred,
+                    then_case.add_context(then_ctx),
+                    else_calse.add_context(else_ctx),
+                ))
+            }
+            Expr::Switch(case_num, branches) => {
+                let new_case_num = case_num.add_context(current_ctx.clone());
+                let new_branches = branches
+                    .iter()
+                    .map(|b| b.clone().add_context(current_ctx.clone()))
+                    .collect();
+                RcExpr::new(Expr::Switch(new_case_num, new_branches))
+            }
+            // for all other nodes, just add the context to the children
+            Expr::Bop(op, x, y) => RcExpr::new(Expr::Bop(
+                op.clone(),
+                x.add_context(current_ctx.clone()),
+                y.add_context(current_ctx),
+            )),
+            Expr::Uop(op, x) => RcExpr::new(Expr::Uop(op.clone(), x.add_context(current_ctx))),
+            Expr::Get(e, i) => RcExpr::new(Expr::Get(e.add_context(current_ctx), *i)),
+            Expr::Alloc(e, ty) => RcExpr::new(Expr::Alloc(e.add_context(current_ctx), ty.clone())),
+            Expr::Call(f, arg) => {
+                RcExpr::new(Expr::Call(f.clone(), arg.add_context(current_ctx.clone())))
+            }
+            Expr::Single(e) => RcExpr::new(Expr::Single(e.add_context(current_ctx))),
+            Expr::Concat(order, x, y) => RcExpr::new(Expr::Concat(
+                order.clone(),
+                x.add_context(current_ctx.clone()),
+                y.add_context(current_ctx),
+            )),
+            // overwrite old context with new, more specific one
+            Expr::InContext(_old_assumption, child) => child.add_context(current_ctx),
```

Comment:
This comment is a bit mysterious:

* When can we have two overlapping InContext nodes?
* Why are later visits more specific?

Reply:
Actually we don't call this method in any places where there is context, so I just made this case panic instead
But the idea was that this pass adds the most specific possible context, since it includes the entire program path.